src/posts/flox-smart/index.md

```diff
@@ -0,0 +1,215 @@
+   Later we will want to merge together the detected cohorts when they occupy _approximately_ the same chunks, using the containment metric.
+1. Now we can quickly determine a number of special cases and exit early:
+   1. Use `"blockwise"` when every group is contained to one block each.
+   1. Use `"cohorts"` when
+      1. every chunk only has a single group, but that group might extend across multiple chunks; and
+      1. existing cohorts don't overlap at all.
+   1. [and more](https://github.com/xarray-contrib/flox/blob/e6159a657c55fa4aeb31bcbcecb341a4849da9fe/flox/core.py#L408-L426)
+
+If we haven't exited yet, then we want to merge together any detected cohorts that substantially overlap with each other using the containment metric.
+
+1. For that we first quickly compute containment for all groups `i` against all other groups `j` as `C = S.T @ S / number_chunks_per_group`.
+1. To choose between `"map-reduce"` and `"cohorts"`, we need a summary measure of the degree to which the labels overlap with
+   each other. We use _sparsity_ --- the number of non-zero elements in `C` divided by the number of elements in `C`, `C.nnz/C.size`.
+   When sparsity is relatively high, we use `"map-reduce"`, otherwise we use `"cohorts"`.
+1. If the sparsity is low enough, we merge together similar cohorts using a for-loop.
+
+For more detail [see the docs](https://flox.readthedocs.io/en/latest/implementation.html#heuristics) or [the code](https://github.com/xarray-contrib/flox/blob/e6159a657c55fa4aeb31bcbcecb341a4849da9fe/flox/core.py#L336).
+Suggestions and improvements are very welcome!
+
+Here is containment `C[i, j]` for a range of chunk sizes from 1 to 12 for computing `groupby("time.month")` of a monthly mean dataset.
+These panels are colored so that light yellow is `C=0`, and dark purple is `C=1`.
+`C[i,j] = 1` when the chunks occupied by group `i` perfectly overlaps with those occupied by group `j` (so the diagonal elements
+are always 1).
+Since there are 12 groups, `C` is a 12x12 matrix.
+The title on each image is `(chunk size, sparsity)`.
+![flox sparsity image](https://flox.readthedocs.io/en/latest/_images/containment.png)
+
+When the chunksize _is_ a divisor of the period 12, `C` is a [block diagonal](https://en.wikipedia.org/wiki/Block_matrix) matrix.
+When the chunksize _is not_ a divisor of the period 12, `C` is much less sparse in comparison.
+Given the above `C`, flox will choose `"cohorts"` for chunk sizes (1, 2, 3, 4, 6), and `"map-reduce"` for the rest.
+
+Cool, isn't it?!
+
+Importantly this inference is fast — [~250ms for the US county](https://flox.readthedocs.io/en/latest/implementation.html#example-spatial-grouping) GroupBy problem in our [previous post](https://xarray.dev/blog/flox) where approximately 3000 groups are distributed over 2500 chunks; and ~1.25s for grouping by US watersheds ~87000 groups across 640 chunks.
+
+## What's next?
```

Comment:
Your conclusion is a bit rambly, and I think some of the content (especially call-outs to any other groupby nerds who can help) could be in a footnote.

I think you should end with a succinct one-line conclusion that emphasises that most of this complexity can now be safely ignored by users.